File: Python3/67.Add_Binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        carry = 0
        m = len(a)
        n = len(b)
        res = ''

        for i in range(max(m,n)):
            a_digit = 0 if i >= m or a[-i-1] == '0' else 1
            b_digit = 0 if i >= n or b[-i-1] == '0' else 1
            total = a_digit + b_digit + carry

            carry = 0 if total < 2 else 1
            res = ('0' if total % 2 == 0 else '1') + res

        return '1' + res if carry else res

File: Java/67.Add_Binary.java
class Solution {
    public String addBinary(String a, String b) {
        int carry = 0;
        int m = a.length();
        int n = b.length();
        String res = "";

        int i = m - 1, j = n - 1;
        while(i >= 0 || j >= 0) {
            int a_bit = 0;
            int b_bit = 0;

            if (i >= 0) {
                a_bit = a.charAt(i--) - '0';
            }
            if (j >= 0) {
                b_bit = b.charAt(j--) - '0';
            }
            res = String.valueOf(a_bit ^ b_bit ^ carry) + res;
            carry = (a_bit + b_bit + carry >= 2) ? 1 : 0;
        }
        return carry == 1 ? "1" + res : res;
    
    }
}